types/mempool/priority.go

```diff
@@ -0,0 +1,363 @@
+}
+
+type PriorityMempoolOption func(*priorityMempool)
+
+// WithOnRead sets a callback to be called when a tx is read from the mempool.
+func WithOnRead(onRead func(tx Tx)) PriorityMempoolOption {
+	return func(mp *priorityMempool) {
+		mp.onRead = onRead
+	}
+}
+
+// DefaultPriorityMempool returns a priorityMempool with no options.
+func DefaultPriorityMempool() Mempool {
+	return NewPriorityMempool()
+}
+
+// NewPriorityMempool returns the SDK's default mempool implementation which
+// returns txs in a partial order by 2 dimensions; priority, and sender-nonce.
+func NewPriorityMempool(opts ...PriorityMempoolOption) Mempool {
+	mp := &priorityMempool{
+		priorityIndex:  huandu.New(huandu.LessThanFunc(txMetaLess)),
+		priorityCounts: make(map[int64]int),
+		senderIndices:  make(map[string]*huandu.SkipList),
+		senderCursors:  make(map[string]*huandu.Element),
+		scores:         make(map[txMeta]txMeta),
+	}
+
+	for _, opt := range opts {
+		opt(mp)
+	}
+
+	return mp
+}
+
+// Insert attempts to insert a Tx into the app-side mempool in O(log n) time,
+// returning an error if unsuccessful. Sender and nonce are derived from the
+// transaction's first signature.
+//
+// Transactions are unique by sender and nonce. Inserting a duplicate tx is an
+// O(log n) no-op.
+//
+// Inserting a duplicate tx with a different priority overwrites the existing tx,
+// changing the total order of the mempool.
+func (mp *priorityMempool) Insert(ctx sdk.Context, tx Tx) error {
+	sigs, err := tx.(signing.SigVerifiableTx).GetSignaturesV2()
+	if err != nil {
+		return err
+	}
+	if len(sigs) == 0 {
+		return fmt.Errorf("tx must have at least one signer")
+	}
+
+	priority := ctx.Priority()
+	sig := sigs[0]
+	sender := sig.PubKey.Address().String()
+	nonce := sig.Sequence
+	key := txMeta{nonce: nonce, priority: priority, sender: sender}
+
+	senderIndex, ok := mp.senderIndices[sender]
+	if !ok {
+		senderIndex = huandu.New(huandu.LessThanFunc(func(a, b any) int {
+			return huandu.Uint64.Compare(b.(txMeta).nonce, a.(txMeta).nonce)
+		}))
+
+		// initialize sender index if not found
+		mp.senderIndices[sender] = senderIndex
+	}
+
+	mp.priorityCounts[priority]++
+
+	// Since senderIndex is scored by nonce, a changed priority will overwrite the
+	// existing key.
+	senderTx := senderIndex.Set(key, tx)
+
+	// Since mp.priorityIndex is scored by priority, then sender, then nonce, a
+	// changed priority will create a new key, so we must remove the old key and
+	// re-insert it to avoid having the same tx with different priorityIndex indexed
+	// twice in the mempool.
+	//
+	// This O(log n) remove operation is rare and only happens when a tx's priority
+	// changes.
+	sk := txMeta{nonce: nonce, sender: sender}
+	if oldScore, txExists := mp.scores[sk]; txExists {
+		mp.priorityIndex.Remove(txMeta{
+			nonce:    nonce,
+			sender:   sender,
+			priority: oldScore.priority,
+			weight:   oldScore.weight,
+		})
+	}
+
+	mp.scores[sk] = txMeta{priority: priority, weight: key.weight}
```

Comment:
The key for `scores` is always `sender-nonce` and the value is either priority or priority and weight (at a non-default value).  Given this I'm not following what you mean by different key space?

`scores` is used to store and lookup a given tx's priority and weight.  Since a tx is uniquely identified by `sender-nonce` that is the key.  We need to lookup priority/weight for delete operations to locate the tx in the priority list (they are part of the key) since we're only given tx (not priority or weight).

Actually looking at this now we don't need to set `weight` here in the scores index (during insert) since I moved to a single pass iteration to calculate scores at select time.

Reply:
Key = `txMeta{nonce: nonce, sender: sender}`
Value = `txMeta{priority: priority, weight: key.weight}` 
?

Reply:
Mhhh not sure if I misread or of it was changed. In any case, resolving 👍